### ISSUE TITLE
grid_sample with align_corner=True and mode='bilinear' induces errors at discrete pixel locations

### ISSUE THREAD

**Issue by tvercaut** [pytorch/pytorch#81868]
### 🐛 Describe the bug

For an tensor `I`, if location `x` and `y` are integers (before normalisation), the result of bilinear interpolation of `I` at `(x,y)` should be strictly equal to `I(x,y)`. This is not currently the case when using `grid_sample` with `align_corners=True` and `mode='bilinear'`.

### Reproduction steps

```python
import torch

img = torch.rand(1, 1, 4, 4)
theta = torch.zeros(1,2,3)
theta[0,0,0] = 1
theta[0,1,1] = 1

align_opt = True
grid = torch.nn.functional.affine_grid(theta, img.shape, align_corners=align_opt)
out = torch.nn.functional.grid_sample(img, grid, mode='bilinear', align_corners=align_opt)
print(f"Residual with align_corners={align_opt}: {torch.linalg.vector_norm(out-img, ord=torch.inf).numpy()}")

align_opt = False
grid = torch.nn.functional.affine_grid(theta, img.shape, align_corners=align_opt)
out = torch.nn.functional.grid_sample(img, grid, mode='bilinear', align_corners=align_opt)
print(f"Residual with align_corners={align_opt}: {torch.linalg.vector_norm(out-img, ord=torch.inf).numpy()}")
```

### Observed behavior

```
Residual with align_corners=True: 1.7881393432617188e-07
Residual with align_corners=False: 0.0
```

### Expected behavior

`out` should be equal to `img` for an identity transformation.

### Additional context

The errors are seen in both CPU and GPU mode. This issue was first observed as part of using kornia:
https://github.com/kornia/kornia/issues/1799

### Versions

- PyTorch Version (e.g., 1.0): 1.12.0+cu113
- OS (e.g., Linux): Google colab
- How you installed PyTorch (`conda`, `pip`, source): Google colab default version
- Build command you used (if compiling from source): N/A
- Python version: 3.7.13 (default, Apr 24 2022, 01:04:09)
- CUDA/cuDNN version: cu113
- GPU models and configuration: Tesla T4
- Any other relevant information: N/A

**Comment by tvercaut** [comment]
A root cause of the loss of precision may be related to the normalisation of the pixel coordinates betwee -1 and 1. Indeed a simple baseline implemenation of bilinear interpolation which does not show any error when working in pixel coordinates also shows discrepancy when normalizing and un-normalizing the grid:
```python
import torch
print(f'Running PyTorch version: {torch.__version__}')

torchdevices = [torch.device('cpu')]
if torch.cuda.is_available():
  torchdevices.append(torch.device('cuda'))
  print('Default GPU is ' + torch.cuda.get_device_name(torch.device('cuda')))

# A baseline implementation of linear resampling
def my_grid_sample(input, grid, align_corners=False):
  if align_corners == 'pixel units':
    # do nothing
    0
  else:
    # Nornalise grid to pixel units
    height = input.shape[2]
    width  = input.shape[3]
    if align_corners == False:
      gw = width/2.
      bw = (width-1.)/2.
      gh = height/2.
      bh = (height-1.)/2.
    elif align_corners == True:
      gw = (width-1.)/2.
      bw = (width-1.)/2.
      gh = (height-1.)/2.
      bh = (height-1.)/2.

    grid[:,:,:,0] = gw*grid[:,:,:,0] + bw
    grid[:,:,:,1] = gh*grid[:,:,:,1] + bh

  floorg = torch.floor(grid).to(torch.long)
  decg = grid - floorg
  ceilg = torch.ceil(grid).to(torch.long)
  resdecg = 1.- decg

  floorg[:,:,:,0] = torch.clamp(floorg[:,:,:,0],0,input.shape[3]-1)
  floorg[:,:,:,1] = torch.clamp(floorg[:,:,:,1],0,input.shape[2]-1)

  ceilg[:,:,:,0] = torch.clamp(ceilg[:,:,:,0],0,input.shape[3]-1)
  ceilg[:,:,:,1] = torch.clamp(ceilg[:,:,:,1],0,input.shape[2]-1)

  inw = input[:,:,floorg[0,:,:,1].flatten(),floorg[0,:,:,0].flatten()]
  wnw = resdecg[0,:,:,1] * resdecg[0,:,:,0]
  ine = input[:,:,floorg[0,:,:,1].flatten(),ceilg[0,:,:,0].flatten()]
  wne = resdecg[0,:,:,1] * decg[0,:,:,0]
  isw = input[:,:,ceilg[0,:,:,1].flatten(),floorg[0,:,:,0].flatten()]
  wsw = decg[0,:,:,1] * resdecg[0,:,:,0]
  ise = input[:,:,ceilg[0,:,:,1].flatten(),ceilg[0,:,:,0].flatten()]
  wse = decg[0,:,:,1] * decg[0,:,:,0]

  vals = inw*wnw.flatten() + ine*wne.flatten() + isw*wsw.flatten() + ise*wse.flatten()

  vals = vals.reshape((1, 1, grid.shape[1], grid.shape[2]))
  return vals

for torchdevice in torchdevices:
  print('\nRunning on ' + str(torchdevice))

  img = torch.rand(1, 1, 40, 30, device=torchdevice)
  theta = torch.zeros(1,2,3, device=torchdevice)
  theta[0,0,0] = 1
  theta[0,1,1] = 1

  align_opts = [True, False]
  for align_opt in align_opts:
    grid = torch.nn.functional.affine_grid(theta, img.shape, align_corners=align_opt)
    out = torch.nn.functional.grid_sample(img, grid, mode='bilinear', align_corners=align_opt)
    print(f"Residual with align_corners={align_opt}: {torch.linalg.vector_norm(out-img, ord=torch.inf).cpu().numpy()}")

    out2 = my_grid_sample(img, grid, align_corners=align_opt)
    print(f"Baseline impl - Residual with align_corners={align_opt}: {torch.linalg.vector_norm(out2-img, ord=torch.inf).cpu().numpy()}")

  hh = torch.arange(start=0., end=img.shape[2], device=torchdevice)
  ww = torch.arange(start=0., end=img.shape[3], device=torchdevice)
  gridh, gridw = torch.meshgrid(hh, ww, indexing='ij')
  grid = torch.stack((gridw[None,...], gridh[None,...]),dim=3)
  out3 = my_grid_sample(img, grid, align_corners='pixel units')
  print(f"Baseline impl in pix units - Residual with align_corners={align_opt}: {torch.linalg.vector_norm(out3-img, ord=torch.inf).cpu().numpy()}")
```
with corresponding output:
```
Running PyTorch version: 1.12.0+cu113
Default GPU is Tesla T4

Running on cpu
Residual with align_corners=True: 2.1383166313171387e-06
Baseline impl - Residual with align_corners=True: 8.940696716308594e-07
Residual with align_corners=False: 2.2649765014648438e-06
Baseline impl - Residual with align_corners=False: 9.5367431640625e-07
Baseline impl in pix units - Residual with align_corners=False: 0.0

Running on cuda
Residual with align_corners=True: 2.998858690261841e-06
Baseline impl - Residual with align_corners=True: 6.556510925292969e-07
Residual with align_corners=False: 1.7881393432617188e-06
Baseline impl - Residual with align_corners=False: 1.2069940567016602e-06
Baseline impl in pix units - Residual with align_corners=False: 0.0
```

This issue may thus hinge upon #36107 (See also #24870) and may add further justification for allowing `grid_sample` to work in absolute coordinates (similar to [`scipy.ndimage.map_coordinates`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.map_coordinates.html))

**Comment by tvercaut** [comment]
Interestingly no such precision issue occur whith `torch.nn.functional.interpolate`. Indeed, running
```python
import torch
print(f'Running PyTorch version: {torch.__version__}')

torchdevices = [torch.device('cpu')]
if torch.cuda.is_available():
  torchdevices.append(torch.device('cuda'))
  print('Default GPU is ' + torch.cuda.get_device_name(torch.device('cuda')))

for torchdevice in torchdevices:
  print('Running on ' + str(torchdevice))

  img = torch.rand(1, 1, 40, 40, device=torchdevice)

  align_opts = [True, False]
  for align_opt in align_opts:
    out = torch.nn.functional.interpolate(img, scale_factor=1., mode='bilinear', align_corners=align_opt )
    print(f"Residual with align_corners={align_opt}: {torch.linalg.vector_norm(out-img, ord=torch.inf).cpu().numpy()}")
```
leads to
```
Running PyTorch version: 1.12.0+cu113
Default GPU is Tesla T4
Running on cpu
Residual with align_corners=True: 0.0
Residual with align_corners=False: 0.0
Running on cuda
Residual with align_corners=True: 0.0
Residual with align_corners=False: 0.0
```

**Comment by cako** [comment]
Seems like this is also a problem for nearest, at least for even-sized tensors.

```python
import torch
import numpy as np

for n in range(5, 11):
    # `input`: [0, 1, ..., n-1]
    input = torch.arange(n, dtype=torch.float64).view(1, 1, 1, n)

    # Grid will try to sample coordinates [0, 1, ..., n-1]
    # which must equal the values in `input`.
    grid = torch.arange(n, dtype=torch.float64).view(1, 1, -1, 1)
    grid = grid * 2 / (n - 1) - 1  # Map to [-1, 1]
    grid = torch.cat((grid, -torch.ones_like(grid)), axis=-1)  # Dummy coordinate at -1
    # Ensure mapping is works
    assert grid[..., 0].min() == -1
    assert grid[..., 0].max() == 1

    interp = torch.nn.functional.grid_sample(
        input, grid, align_corners=False, mode="nearest", padding_mode="zeros"
    )
    input_np = input.squeeze().numpy()
    interp_np = interp.squeeze().numpy()
    equal = np.allclose(input_np, interp_np)
    print(f"{n}:", equal)
    print(input_np, interp_np)
    print()
```

**Comment by artths** [comment]
My model goes from being top performer when using grid_sample with "bicubic" mode to failing to train when using "bilinear" no matter how hard I try. Same result when I change align_corners=True to False. I have a feeling something is wrong inside PyTorch, but I don't know where to look at.
